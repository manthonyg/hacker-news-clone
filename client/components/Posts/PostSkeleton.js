import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";

const PostCard = styled.div`
  display: flex;
  align-content: center;
  width: 100%;
  min-height: 100px;
  height: 100%;
  background-color: ${(props) => {
    if (props.theme.theme === "light") return "#FFFFFF";
    return "#282828";
  }};
  border-bottom: ${(props) => {
    if (props.theme.theme === "light") return "1px solid #c3c3c3";
    return "1px solid #404040";
  }};
`;

const withEmoji = css`
  content: "";
  position: absolute;
  border-radius: 100%;
  left: -50px;

  width: 40px;
  height: 40px;
  background-color: ${(props) => {
    if (props.theme.theme === "light") return "#eeeeee";
    return "#212121";
  }};
`;

const Skeleton = styled.div`
  display: flex;
  padding-left: 80px;
  width: 50vw;
`;

const SkeletonContent = styled.span`
  height: 13px;
  position: absolute;
  top: 40px;
  width: 100%;
  background-color: ${(props) => {
    if (props.theme.theme === "light") return "#eeeeee";
    return "#212121";
  }};

  &:nth-child(1):before {
    ${(props) => (props.noEmoji ? "" : withEmoji)}
  }
  &:nth-child(2) {
    height: 8px;
    top: 60px;
  }
`;

const Loading = styled.div`
  position: relative;
  width: 100%;
`;

const SkeletonInner = styled.div`
  height: 75px;
  margin-top: 20px;
`;

function PostSkeleton({ numberOfSkeletons = 1, noEmoji }) {
  return (
    <>
      {[...Array(numberOfSkeletons)].map((skeleton, i) => (
        // eslint-disable-next-line react/no-array-index-key
        <PostCard key={i + numberOfSkeletons}>
          <Skeleton key={i}>
            <Loading>
              <SkeletonContent noEmoji={noEmoji} />
              <SkeletonContent />
              <SkeletonInner />
            </Loading>
          </Skeleton>
        </PostCard>
      ))}
    </>
  );
}

PostSkeleton.propTypes = {
  numberOfSkeletons: PropTypes.number.isRequired,
  noEmoji: PropTypes.bool,
};

export default PostSkeleton;
