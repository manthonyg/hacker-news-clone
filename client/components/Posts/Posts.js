import React, { useState, useEffect, useRef } from "react";
import PostList from "../PostList/PostList";
import { fetchMainPosts, fetchPostIds, fetchPosts } from "../../utils/api";
import Loader from "../common/Loader/Loader";
import InfiniteScroll from "react-infinite-scroll-component";
import Heading from "../common/Heading/Heading";
import Header from "../common/Heading/Heading";
import PropTypes from "prop-types";
function Posts(props) {
  const { type } = props;
  const [isLoading, setIsLoading] = useState(false);
  const [posts, setPosts] = useState([]);
  const [postIds, setPostIds] = useState([]);
  const INFINITE_SCROLL_FETCH_AMOUNT = 10;

  const componentIsMounted = useRef(true);

  useEffect(() => {
    if (componentIsMounted.current) {
      setIsLoading(true);
      fetchMainPosts(type).then((response) => {
        setPosts(response);
        setIsLoading(false);
      });

      fetchPostIds(type)
        .then((response) => setPostIds(response))
        .catch((error) => console.log(error));
    }
    return () => (componentIsMounted.current = false);
  }, [type]);

  const handleInfiniteScroll = () => {
    const currentPostLength = posts.length;
    fetchPosts(
      postIds.slice(
        currentPostLength - 1,
        currentPostLength + INFINITE_SCROLL_FETCH_AMOUNT
      )
    ).then((response) => setPosts(posts.concat(response)));
  };

  return (
    <>
      {isLoading ? (
        <>
          <Header h5>Loading {type}...</Header>
          <Loader />
        </>
      ) : (
        <InfiniteScroll
          style={{ overflow: "auto" }}
          dataLength={posts.length} //This is important field to render the next data
          next={handleInfiniteScroll}
          hasMore={posts?.length < postIds?.length}
          scrollThreshold={0.95}
          loader={
            <>
              <Header h5>Loading {type}...</Header>
              <Loader />
            </>
          }
          endMessage={<Heading>No more posts!</Heading>}
        >
          <PostList posts={posts} />
        </InfiniteScroll>
      )}
    </>
  );
}

Posts.propTypes = {
  type: PropTypes.string.isRequired,
};

export default Posts;
