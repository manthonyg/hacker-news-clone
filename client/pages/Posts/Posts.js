import React, { useState, useEffect, useRef } from "react";
import PostList from "../../components/PostList/PostList";
import { fetchMainPosts, fetchPostIds, fetchPosts } from "../../utils/api";
import Loader from "../../components/common/Loader/Loader";
import InfiniteScroll from "react-infinite-scroll-component";
import Heading from "../../components/common/Heading/Heading";
import Header from "../../components/common/Heading/Heading";
import PropTypes from "prop-types";
import {
  POST_CATEGORY_TEXT,
  POST_LOADER_MESSAGE,
  POST_END_MESSAGE,
} from "../../test_utils/testIds";
import ScrollArrow from "../../components/ScrollArrow/ScrollArrow";
import useInfiniteScroll from "../../hooks/useInfiniteScroll";

function Posts(props) {
  const { type } = props;
  const [isLoading, setIsLoading] = useState(false);
  const [posts, setPosts] = useState([]);
  const [postIds, setPostIds] = useState([]);
  const INFINITE_SCROLL_FETCH_AMOUNT = 10;
  const [isFetching, setIsFetching] = useInfiniteScroll(() =>
    handleInfiniteScroll()
  );
  const [hasPosts, setHasPosts] = useState(true);
  const componentIsMounted = useRef(true);

  useEffect(() => {
    if (componentIsMounted.current) {
      setIsLoading(true);
      fetchMainPosts(type).then((response) => {
        setPosts(response);
        setIsLoading(false);
      });

      fetchPostIds(type)
        .then((response) => setPostIds(response))
        .catch((error) => console.log(error));
    }
    return () => (componentIsMounted.current = true);
  }, [props.type]);

  const checkIfHasPosts = () => {
    console.log("checking for posts");
    if (posts.length >= postIds.length) {
      setHasPosts(false);
    }
  };

  const handleInfiniteScroll = () => {
    if (hasPosts) {
      setTimeout(() => {
        const currentPostLength = posts.length;
        fetchPosts(
          postIds.slice(
            currentPostLength,
            currentPostLength + INFINITE_SCROLL_FETCH_AMOUNT + 1
          )
        ).then((response) => {
          setPosts(posts.concat(response));
          setIsFetching(false);
          checkIfHasPosts();
        });
      }, 200);
    }
  };

  return (
    <>
      {isLoading ? (
        <>
          <Header data-testid={POST_LOADER_MESSAGE} h5>
            loading {type}...
          </Header>
          <Loader />
        </>
      ) : (
        <>
          <Heading data-testid={POST_CATEGORY_TEXT} isSticky h5>
            viewing {type} ({posts.length} of {postIds.length})
          </Heading>
          <PostList posts={posts} category={type} />

          {isFetching && hasPosts && (
            <>
              <Heading h5 data-testid={POST_LOADER_MESSAGE}>
                loading {type}...
              </Heading>
              <Loader />
            </>
          )}

          {!hasPosts && (
            <Heading h4 data-testid={POST_END_MESSAGE}>
              no more posts
            </Heading>
          )}
          <ScrollArrow />
        </>
      )}
    </>
  );
}

Posts.propTypes = {
  type: PropTypes.string.isRequired,
};

export default Posts;
