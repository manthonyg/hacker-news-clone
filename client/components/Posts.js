/* eslint-disable no-console */
import React from 'react';
import PostList from './PostList';
import { fetchMainPosts } from '../utils/api';
import '../index.css';
import PostSkeleton from './PostSkeleton';

export default class Posts extends React.Component {
  constructor(props) {
    super(props);

    this.state = { isLoading: false, posts: [] };
    this.handleFetch = this.handleFetch.bind(this);
  }

  componentDidMount() {
    this.setState({ isLoading: true });
    this.handleFetch();
  }

  componentDidUpdate(previousProps) {
    const { type } = this.props;
    if (previousProps.type !== type) {
      this.handleFetch();
    }
  }

  handleFetch() {
    const { type } = this.props;
    this.setState({ isLoading: true, posts: null });
    fetchMainPosts(type).then(data => {
      console.log(data);

      this.setState({ posts: data, isLoading: false });
    });
  }

  render() {
    const { isLoading, posts } = this.state;
    return (
      <>
        {isLoading && (
          <>
            <PostSkeleton numberOfSkeletons={20} />
          </>
        )}

        <PostList posts={posts} isLoading={isLoading} />
      </>
    );
  }
}
