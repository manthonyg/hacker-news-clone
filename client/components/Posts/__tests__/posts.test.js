import React from "react";
import renderer from "react-test-renderer";
import Posts from "../../Posts/Posts";

it("renders correctly with all props", () => {
  const requiredProps = {
    type: "new",
  };

  const tree = renderer
    .create(
      <>
        <Posts {...requiredProps} />
      </>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
