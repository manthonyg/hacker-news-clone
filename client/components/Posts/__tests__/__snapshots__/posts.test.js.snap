// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<div
  className="infinite-scroll-component__outerdiv"
  style={Object {}}
>
  <div
    className="infinite-scroll-component "
    style={
      Object {
        "WebkitOverflowScrolling": "touch",
        "height": "auto",
        "overflow": "auto",
      }
    }
  >
    <h1
      className="sc-fzozJi kNFnqJ"
    >
      No more posts!
    </h1>
  </div>
</div>
`;
